lib/promscrape/discovery/yandexcloud/api.go

```diff
@@ -0,0 +1,272 @@
+package yandexcloud
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promauth"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discoveryutils"
+)
+
+const (
+	defaultInstanceCredsEndpoint = "http://169.254.169.254/latest/meta-data/iam/security-credentials/default"
+	defaultAPIEndpoint           = "https://api.cloud.yandex.net"
+	defaultAPIVersion            = "v1"
+)
+
+var configMap = discoveryutils.NewConfigMap()
+
+type apiCredentials struct {
+	Token      string    `json:"Token"`
+	Expiration time.Time `json:"Expiration"`
+}
+
+// yandexPassportOAuth is a struct for Yandex Cloud IAM token request
+// https://cloud.yandex.com/en-ru/docs/iam/operations/iam-token/create
+type yandexPassportOAuth struct {
+	YandexPassportOAuthToken string `json:"yandexPassportOauthToken"`
+}
+
+// iamToken Yandex Cloud IAM token response
+// https://cloud.yandex.com/en-ru/docs/iam/operations/iam-token/create
+type iamToken struct {
+	IAMToken  string    `json:"iamToken"`
+	ExpiresAt time.Time `json:"expiresAt"`
+}
+
+type apiConfig struct {
+	client              *http.Client
+	tokenLock           sync.Mutex
+	creds               *apiCredentials
+	yandexPassportOAuth *yandexPassportOAuth
+	serviceEndpoints    map[string]*url.URL
+}
+
+func getAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	v, err := configMap.Get(sdc, func() (interface{}, error) { return newAPIConfig(sdc, baseDir) })
+	if err != nil {
+		return nil, err
+	}
+	return v.(*apiConfig), nil
+}
+
+func newAPIConfig(sdc *SDConfig, baseDir string) (*apiConfig, error) {
+	cfg := &apiConfig{
+		client: &http.Client{
```

Comment:
Probably it would be better using the [discoveryutils.NewClient](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/5d364545bda670c03faab6de14400174d633cc6a/lib/promscrape/discoveryutils/client.go#L62) here if possible.

Reply:
Hi @valyala, thank you for the feedback. I need some time to investigate the possibility. m.b. you know is there any way to change the config of such client on the fly? yc api requires time limited bearer token and there `getFreshAPICredentials()` for the update

Reply:
I've found a way in Azure SD, I'll implement if it won't complicate the code.

Reply:
@valyala Hi, YC API uses per-service [edpoints](https://cloud.yandex.com/en-ru/docs/api-design-guide/concepts/endpoints), so it is difficult to use one [discoveryutils.NewClient](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/5d364545bda670c03faab6de14400174d633cc6a/lib/promscrape/discoveryutils/client.go#L62) for scraping, isn't it ?

Reply:
Agreed. Let's leave standard `http.Client` then here.